ISSUE #1561: Fixes create and restore database with drush arr

sherakama opened the pull request:
Suggested fix for issue #1561

weitzman commented on lib/Drush/Sql/SqlBase.php:
```diff
@@ -257,7 +257,10 @@ public function drop_or_create() {
+      $original_spec = $this->db_spec;
       $this->createdb();
+      // Restore the original spec as create db and su() can modify the spec.
+      $this->db_spec = $original_spec;
```
I didnt investigate, but it is a bit odd to wrap createdb here. Can this dance not be done inside createdb()?

sherakama replied:
That might be a better approach. I'll take a look to see.

weitzman commented on lib/Drush/Sql/Sqlmysql.php:
```diff
@@ -32,7 +32,9 @@ public function creds($hide_password = TRUE) {
     // Some drush commands (e.g. site-install) want to connect to the
     // server, but not the database.  Connect to the built-in database.
-    $parameters['database'] = empty($this->db_spec['database']) ? 'information_schema' : $this->db_spec['database'];
+    if (!empty($this->db_spec['database'])) {
```
I'm a little hesitant to remove information_schema. I think we need to specify a DB in certain cases (see early parts of site-install command). Needs investigation.

sherakama replied:
information_schema cannot be written to. Read is fine, but I wonder if we can protect this from defaulting to information_schema when it is a write operation.

weitzman replied:
In that case the caller should have provided a database. I'm not too concerned about this.

sherakama commented:
Ok, moved the db_spec restore to the createdb method and added a flag for when creating a new database to avoid adding the --database parameter to the mysql statement. 

Debug output now looks like:

```
Executing: mysql --defaults-extra-file=/private/tmp/drush_OlVRJz --database=mydatabase --host=localhost --silent  < /private/tmp/drush_L8XnDW > /dev/null

Executing: mysql --defaults-extra-file=/private/tmp/drush_dgBIe0 --host=localhost --silent  < /private/tmp/drush_NnPnGb

Executing: mysql --defaults-extra-file=/private/tmp/drush_YEIDG2 --database=mydatabase --host=localhost --silent  < /tmp/drush_tmp_1440615699_55de0d13b289c/ds_jumpstart_cars.sql
```

You can see that the second mysql statement does not include the database parameter as that one has the create new db statement in it.

weitzman commented on lib/Drush/Sql/SqlBase.php:
```diff
@@ -239,11 +239,26 @@ public function createdb_sql($dbname, $quoted = FALSE) {}
+    $this->db_spec = $original_spec;
+
+    // Return the result of the query.
+    return $result;
```
This hunk looks good to me.

weitzman commented on lib/Drush/Sql/Sqlmysql.php:
```diff
@@ -34,6 +34,12 @@ public function creds($hide_password = TRUE) {
+    if (isset($this->db_spec['new-database']) && $this->db_spec['new-database']) {
+      unset($parameters['database']);
+    }
+
```
Maybe creds() should always return a full array and it is up to the caller to remove 'database' if needed?

sherakama replied:
Outside of creating or dropping a database I don't see a need to remove the database option. Perhaps changing the key from 'new-database' to 'no-database' or something similar?

sherakama commented:
Moved this to: https://github.com/drush-ops/drush/pull/1872
